Override `HttpRequestDecoder.createMessage()` for perfomance

echo304 opened the pull request:
Motivation:

By overriding `HttpRequestDecoder.createMessage()`, performance can be improved because there is no need to copy original netty header.

Modifications:

- TBD

Result:

- Closes #4853 
- Performance is improved

<!--
Visit this URL to learn more about how to write a pull request description:
https://armeria.dev/community/developer-guide#how-to-write-pull-request-description
-->

jrhee17 commented on core/src/main/java/com/linecorp/armeria/server/HttpServerCodec.java:
```diff
@@ -0,0 +1,171 @@
+package com.linecorp.armeria.server;
```
Missing copyright https://armeria.dev/community/developer-guide#add-copyright-header

Also, can you copy over the original license?
https://github.com/line/armeria/blob/1def296db95a8d5a2dfdf71600e621db5c649f7a/core/src/main/java/com/linecorp/armeria/server/Http2ServerUpgradeCodec.java#L16-L29

jrhee17 commented on core/src/main/java/com/linecorp/armeria/server/HttpServerCodec.java:
```diff
@@ -0,0 +1,171 @@
+import io.netty.handler.codec.http.HttpServerUpgradeHandler;
+import io.netty.handler.codec.http.HttpStatusClass;
+import io.netty.handler.codec.http.HttpVersion;
+
```
Can you leave the permalink of the forked file in the comments?

https://github.com/line/armeria/blob/1def296db95a8d5a2dfdf71600e621db5c649f7a/core/src/main/java/com/linecorp/armeria/server/Http2ServerUpgradeCodec.java#L65

jrhee17 reviewed (commented):
Thanks! Left some preliminary comments

echo304 commented:
@jrhee17 This PR is not ready to be reviewed. Sorry about that 🙇 
I changed it to draft PR!

trustin commented:
Thanks @echo304 for this 🙇 Looking forward to some performance boost!

echo304 commented on core/src/main/java/com/linecorp/armeria/server/HttpServerCodec.java:
```diff
@@ -0,0 +1,203 @@
+
+        @Override
+        protected HttpMessage createMessage(String[] initialLine) throws Exception {
+            return new DefaultHttpRequest(
```
After some investigation, I found the logic that converts Netty request to Armeria headers.
https://github.com/line/armeria/blob/92e70170b560470a68ae18895d60b3689c8d98bb/core/src/main/java/com/linecorp/armeria/internal/common/ArmeriaHttpUtil.java#L627-L649

At the very beginning, I thought that we can implements `io.netty.handler.codec.http.HttpRequest` and create our own HttpRequest implementation to replace `DefaultHttpRequest`.
But with this design, we still need `ArmeriaHttpUtil.toArmeria` some part of the logic because this class(HttpServerCodec) depends on netty's interface.

minwoox replied:
>  I thought that we can implements io.netty.handler.codec.http.HttpRequest and create our own HttpRequest implementation to replace DefaultHttpRequest.

I think it's the way to go. 👍  We can probably create `RequestHeadersBuilder` in it and passes it in `Http1RequestDecoder`.

> we still need ArmeriaHttpUtil.toArmeria some part of the logic because this class(HttpServerCodec) depends on netty's interface.

Yeah, we can apply the logic at once in the `Http1RequestDecoder` or we can just use the logic inside of the `HttpHeaders` that is returned by `HttpMessage.headers()`. 

Does this make sense? 🤔

echo304 replied:
Just want to double check. ✅ 

Is it what we want to improve/replace?
https://github.com/line/armeria/blob/361c115e2a8699bbe3f8d97391df3b37f9330b71/core/src/main/java/com/linecorp/armeria/server/Http1RequestDecoder.java#L173-L175

minwoox replied:
Yes, exactly. 👍 
We can define our own `DefaultHttpRequest` and return it in `createMessage`:
```java
@Override
protected HttpMessage createMessage(String[] initialLine) throws Exception {
    return new ArmeriaDefaultHttpRequest(HttpMethod.valueOf(initialLine[0]), initialLine[1]);
}

static final class ArmeriaDefaultHttpRequest implements HttpRequest {

    private final HttpMethod method;
    private final RequestHeadersBuilder builder;
    private final HttpHeaders headers;

    ArmeriaDefaultHttpRequest(HttpMethod method, String uri) {
        this.method = method;
        builder = RequestHeaders.builder(
                com.linecorp.armeria.common.HttpMethod.valueOf(method.name()), uri);
        headers = new ArmeriaHttpHeaders(builder);
    }

    @Override
    public HttpMethod method() {
        return method;
    }

    @Override
    public HttpHeaders headers() {
        return headers;
    }
    ...
}
```
We also need `ArmeriaHttpHeaders` class and it will prevents adding not allowed headers:
```java
 final class ArmeriaHttpHeaders extends HttpHeaders {

    private final RequestHeadersBuilder builder;

    ArmeriaHttpHeaders(RequestHeadersBuilder builder) {
        this.builder = builder;
    }

    @Override
    public String get(String name) {
        return builder.get(name);
    }

    @Override
    public Integer getInt(CharSequence name) {
        return builder.getInt(name);
    }

    @Override
    public HttpHeaders add(String name, Object value) {
        // We can add a logic that prevents adding invalid headers that is defined in ArmeriaHttpUtil.toArmeria()
        return builder.add(name, value);
    }
    ...
}
```

In `Http1RequestDecoder`, we will bring the `RequestHeadersBuilder` from the Netty request and add a host header if needed:
```java
if (nettyReq instanceof ArmeriaDefaultHttpRequest) {
    final RequestHeadersBuilder builder =
            ((ArmeriaDefaultHttpRequest) nettyReq).requestHeadersBuilder();
}
```

This is what I have imagined. 😉
Please let me know if it works or not.

minwoox replied:
> Is it correct?

That's correct. 😄 

> But my concern is that implementing own ArmeriaHttpHeaders is pretty huge task. (Well, obviously this is unavoidable)

Yes, it is. But we don't have to implement all of the methods. It's only used in the pipeline so we can implement the methods that are called in the handler. 😉

echo304 replied:
🙆‍♂️ 
Let me show you some code as soon as possible even if it's not completed!

trustin commented:
Gentle ping, @echo304. Looking forward to this improvement! 🤞

minwoox commented on core/src/main/java/com/linecorp/armeria/server/Http1RequestDecoder.java:
```diff
@@ -171,8 +173,30 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
-                    final RequestHeaders headers =
-                            ArmeriaHttpUtil.toArmeria(ctx, nettyReq, cfg, scheme.toString(), reqTarget);
+                    final RequestHeaders headers;
+                    if (nettyReq instanceof ArmeriaDefaultHttpRequest) {
```
We could probably do
`assert nettyReq instanceof ArmeriaDefaultHttpRequest`

minwoox commented on core/src/main/java/com/linecorp/armeria/server/Http1RequestDecoder.java:
```diff
@@ -171,8 +173,30 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
+                    if (nettyReq instanceof ArmeriaDefaultHttpRequest) {
+                        final RequestHeadersBuilder builder =
+                                ((ArmeriaDefaultHttpRequest) nettyReq).requestHeadersBuilder();
+
+                        final ArmeriaDefaultHttpRequest armeriaDefaultHttpRequest = (ArmeriaDefaultHttpRequest) nettyReq;
+                        final HttpHeaders inHeaders = armeriaDefaultHttpRequest.headers();
+
+                        builder.sizeHint(inHeaders.size());
```
We don't need to set the sizeHint:
```suggestion
```

minwoox commented on core/src/main/java/com/linecorp/armeria/server/HttpServerCodec.java:
```diff
@@ -0,0 +1,202 @@
+import io.netty.handler.codec.http.HttpStatusClass;
+import io.netty.handler.codec.http.HttpVersion;
+
+public final class HttpServerCodec extends CombinedChannelDuplexHandler<HttpRequestDecoder, HttpResponseEncoder>
```
Could you use this class instead of Netty's one from here?
https://github.com/line/armeria/blob/b78d9515dc8b4b997dd65787f717443b38a21869/core/src/main/java/com/linecorp/armeria/server/WebSocketSessionHandler.java#L127

minwoox commented on core/src/main/java/com/linecorp/armeria/internal/common/ArmeriaHttpHeaders.java:
```diff
@@ -0,0 +1,199 @@
+package com.linecorp.armeria.internal.common;
```
missing copyright

minwoox commented on core/src/main/java/com/linecorp/armeria/internal/common/ArmeriaHttpHeaders.java:
```diff
@@ -0,0 +1,199 @@
+import io.netty.handler.codec.http.HttpHeaders;
+import io.netty.util.AsciiString;
+
+public final class ArmeriaHttpHeaders extends HttpHeaders {
```
We can put this method in the package `com.linecorp.armeria.server` and remove the public modifier.

minwoox commented on core/src/main/java/com/linecorp/armeria/internal/common/ArmeriaHttpHeaders.java:
```diff
@@ -0,0 +1,199 @@
+        }
+
+        if (asciiName.equals(HttpHeaderNames.COOKIE)) {
+            final StringJoiner cookieJoiner = new StringJoiner(COOKIE_SEPARATOR);
```
It seems like `cookieJoiner` is not used.

minwoox commented on core/src/main/java/com/linecorp/armeria/internal/common/ArmeriaHttpHeaders.java:
```diff
@@ -0,0 +1,199 @@
+        builder.clear();
+        return this;
+    }
+}
```
missing new line

minwoox commented on core/src/main/java/com/linecorp/armeria/internal/common/ArmeriaHttpHeaders.java:
```diff
@@ -0,0 +1,199 @@
+    }
+
+    @Override
+    public HttpHeaders add(String name, Object value) {
```
Could you create a test for this class?
We can move the tests for converting Netty HTTP/1.1 to Armeria headers in `ArmeriaHttpUtilTest`

echo304 replied:
Sure I will!

minwoox commented on core/src/main/java/com/linecorp/armeria/server/ArmeriaDefaultHttpRequest.java:
```diff
@@ -0,0 +1,138 @@
+package com.linecorp.armeria.server;
```
missing copyright

minwoox commented on core/src/main/java/com/linecorp/armeria/server/ArmeriaDefaultHttpRequest.java:
```diff
@@ -0,0 +1,138 @@
+import io.netty.handler.codec.http.HttpVersion;
+import io.netty.util.internal.ObjectUtil;
+
+public class ArmeriaDefaultHttpRequest extends DefaultHttpMessage implements HttpRequest {
```
```suggestion
final class ArmeriaDefaultHttpRequest extends DefaultHttpMessage implements HttpRequest {
```
We can remove the public modifiers in this class.

minwoox commented on core/src/main/java/com/linecorp/armeria/server/ArmeriaDefaultHttpRequest.java:
```diff
@@ -0,0 +1,138 @@
+     * @param method      the HTTP method of the request
+     * @param uri         the URI or path of the request
+     */
+    public ArmeriaDefaultHttpRequest(HttpVersion httpVersion, HttpMethod method, String uri) {
```
Let's remove the constructor which isn't used.

minwoox commented on core/src/main/java/com/linecorp/armeria/server/ArmeriaDefaultHttpRequest.java:
```diff
@@ -0,0 +1,138 @@
+     * @param uri               the URI or path of the request
+     * @param givenHeaders      the Headers for this Request
+     */
+    public ArmeriaDefaultHttpRequest(HttpVersion httpVersion, HttpMethod method, String uri, HttpHeaders givenHeaders) {
```
ditto

minwoox commented on core/src/main/java/com/linecorp/armeria/server/ArmeriaDefaultHttpRequest.java:
```diff
@@ -0,0 +1,138 @@
+    }
+
+//    @Override
+//    public String toString() {
```
We can use `MoreObjects.toStringHelper()` for `toString()`.

minwoox reviewed (commented):
Looks good overall. 😉

echo304 commented on core/src/main/java/com/linecorp/armeria/server/ArmeriaHttpHeaders.java:
```diff
@@ -0,0 +1,252 @@
+            splitByCommaAndAdd(connectionDisallowedList, lowerCased);
+        }
+
+        if (HTTP_TO_HTTP2_HEADER_DISALLOWED_LIST.contains(asciiName) ||
```
With this line of code, this headers won't contain `Transfer-Encoding` header.
As a result, below test failed because it considers corresponding request as if `endOfStream`.
<img width="1277" alt="스크린샷 2023-06-08 오후 8 51 41" src="https://github.com/line/armeria/assets/16456651/9d6b7627-b8f6-46de-8844-6bb1a3b3e683">

https://github.com/line/armeria/blob/b808c3e7fb7deb604e46c2c82c189a30ed7822f4/core/src/main/java/com/linecorp/armeria/server/Http1RequestDecoder.java#L262

So I'm trying to solve these kind of regression without copying another header.

minwoox replied:
Oops, I missed this comment. If so, I think we can remove the `Transfer-Encoding` header in `Http1RequestDecoder` after `isTransferEncodingChunked` is executed. WDYT?

echo304 replied:
@minwoox 
I tried another approach that moved `isTransferEncodingChunked` up before filtering out disallowed headers.
https://github.com/line/armeria/pull/4863/commits/a2a2f4853f1bb4bfc48b8ff2d1ddcf6ec7b3cd5e#diff-a3306f0289f199b1657bb0e32849fd97502c82a91318adc373f8e750a96f84ecR187-R188

echo304 commented:
I put all the logics about processing headers in `Http1RequestDecoder` in https://github.com/line/armeria/pull/4863/commits/a2a2f4853f1bb4bfc48b8ff2d1ddcf6ec7b3cd5e
(And fixed a bug in https://github.com/line/armeria/pull/4863/commits/deaa03c4b137d8a7f86d9d7a3803673d701dda45)

After that I realized `Http1RequestDecoder` has too many logic so I moved those ☝️ logics into `ArmeriaHttpHeaders.buildRequestHeaders` in https://github.com/line/armeria/pull/4863/commits/665c104d17d6738835806b7b07aba60595587123

But it still has some failed test should be fixed 😢

minwoox commented:
> But it still has some failed test should be fixed

Please, let me know if you find any trouble while fixing it. 😄

echo304 commented:
With current version, the of `com.linecorp.armeria.server.ServerTest#testUnsupportedMethod` does fail with below result.
<img width="336" alt="스크린샷 2023-07-03 오후 2 45 33" src="https://github.com/line/armeria/assets/16456651/f79c134a-8704-4636-a887-c2953c2f990a">

After some investigation, I found that `msg` of `com.linecorp.armeria.server.Http1RequestDecoder#channelRead` method is `DefaultFullHttpRequest` not `ArmeriaDefaultHttpRequest`.
<img width="1519" alt="스크린샷 2023-07-03 오후 2 43 53" src="https://github.com/line/armeria/assets/16456651/9eda5f7c-2f28-4f74-a1f5-247aec4eb237">

 But I couldn't figure out _**1)why the request is mapped with `DefaultFullHttpRequest`**_ and **_2)the location where it's mapped._**

@minwoox Can you provide some advice regarding it?

minwoox commented on core/src/main/java/com/linecorp/armeria/server/ArmeriaDefaultHttpRequest.java:
```diff
@@ -0,0 +1,144 @@
+        super(httpVersion, validateHeaders, false);
+        this.method = checkNotNull(method, "method");
+        this.uri = checkNotNull(uri, "uri");
+        builder = RequestHeaders.builder(
+                com.linecorp.armeria.common.HttpMethod.valueOf(method.name()), uri);
```
> But I couldn't figure out 1)why the request is mapped with DefaultFullHttpRequest and 2)the location where it's mapped.
@minwoox Can you provide some advice regarding it?

It seems like an exception is raised while converting the method. We need to fix it:
```suggestion
        final com.linecorp.armeria.common.HttpMethod armeriaMethod =
                com.linecorp.armeria.common.HttpMethod.tryParse(method.name());
        builder = RequestHeaders.builder(armeriaMethod != null ? armeriaMethod
                                                               : com.linecorp.armeria.common.HttpMethod.UNKNOWN,
                                         uri);
```

ikhoon commented on core/src/main/java/com/linecorp/armeria/server/ArmeriaDefaultHttpRequest.java:
```diff
@@ -0,0 +1,147 @@
+import io.netty.handler.codec.http.HttpVersion;
+import io.netty.util.internal.ObjectUtil;
+
+final class ArmeriaDefaultHttpRequest extends DefaultHttpMessage implements HttpRequest {
```
Is there a specific reason to extend `DefaultHttpMessage`? That internally creates `DefaultHttpHeaders` again. Duplicate headers may result in performance degression. 
```suggestion
final class ArmeriaHttp1Request implements HttpRequest {
```

ikhoon commented on core/src/main/java/com/linecorp/armeria/server/ArmeriaHttpHeaders.java:
```diff
@@ -0,0 +1,297 @@
+     */
+    public ArmeriaHttpHeaders(RequestHeadersBuilder builder, HttpHeaders headers) {
+        this.builder = builder;
+        headers.forEach(e -> add(e.getKey(), e.getValue()));
```
This copies the Netty headers to Armeria Headers. Unfortunately, it does not match with the goal of this PR.
> there is no need to copy original netty header.

What do you think of lazily creating or mutating fields when necessary?

```java
private final HttpHeaders nettyHeaders;
@Nullable 
private RequestHeaders armeriaHeaders; 

public ArmeriaHttpHeaders(RequestHeadersBuilder builder, HttpHeaders headers) {
   this.headers = headers;
}

@Override
public String get(String name) {
    if (armeriaHeaders != null) {
        return armeriaHeaders.get(name);
    }
    if (needsValidation(name)) {
        // Create Armeria headers and return;
    }
    // Use Netty Headers as is.
    return headers.get(name);
}

public List<Entry<String, String>> entries() {
    // Create Armeria Headers if null
}
public Iterator<Entry<String, String>> iterator() {
    // Create Armeria Headers if null
}
```

echo304 replied:
Thank you for pointing out!
Let me take a look into it :)

echo304 replied:
I've reviewed your idea and bring some thoughts.

Firstly, I totally agree that current approach doesn't resolve the problem of copying headers 🤦 

And specifically against your advice, `Http1RequestDecoder` has plenty of logic that utilize `RequestHeaders`. So I think we need to convert(create) `HttpHeaders` to `RequestHeaders` most of the time even with 'lazy initializing' strategy. Which means it goes back to copying headers situation again.
I'm not sure if 'lazy initialization' strategy could provide significant value. 🤔 **Correct me if I'm wrong.**

With above idea, I'm afraid we need to copy headers somewhere in our system anyway.

> This copies the Netty headers to Armeria Headers. Unfortunately, it does not match with the goal of this PR.
> 
> > there is no need to copy original netty header.
> 
> What do you think of lazily creating or mutating fields when necessary?
> 
> ```java
> private final HttpHeaders nettyHeaders;
> @Nullable 
> private RequestHeaders armeriaHeaders; 
> 
> public ArmeriaHttpHeaders(RequestHeadersBuilder builder, HttpHeaders headers) {
>    this.headers = headers;
> }
> 
> @Override
> public String get(String name) {
>     if (armeriaHeaders != null) {
>         return armeriaHeaders.get(name);
>     }
>     if (needsValidation(name)) {
>         // Create Armeria headers and return;
>     }
>     // Use Netty Headers as is.
>     return headers.get(name);
> }
> 
> public List<Entry<String, String>> entries() {
>     // Create Armeria Headers if null
> }
> public Iterator<Entry<String, String>> iterator() {
>     // Create Armeria Headers if null
> }
> ```

jrhee17 commented:
This PR has been stale for a while - I've added a commit to get it moving forward. Changes include:

- Now, only a single header map is created, followed by a "purge" of http1 only headers
  - This design is necessary because netty/armeria assumes the `nettyReq` still has a `CONNECTION` header in some places. Hence, the purge is done when the netty request is transformed to an armeria request. Note that from this time, http1-only headers won't be available.
  - The design depends on the fact that there are less HTTP1-only headers than the overall headers (which I think should be the majority of the case anyways)
- Renamed the classes from `ArmeriaHttpHeaders`, `ArmeriaDefaultHttpRequest` to `NettyHttp1Headers`, `NettyHttp1Request`
  - Also introduced `ArmeriaHttpHeadersFactory.java` to save more allocation
- Cleaned up some code

Let me know if anything doesn't make sense 🙇

jrhee17 commented on core/src/main/java/com/linecorp/armeria/internal/common/ArmeriaHttpUtil.java:
```diff
@@ -696,6 +693,8 @@ public static HttpHeaders toArmeria(io.netty.handler.codec.http.HttpHeaders inHe
      * Converts the specified Netty HTTP/1 headers into Armeria HTTP/2 headers.
+     * Functionally, this method is expected to behavior in the same way as
+     * {@link #purgeHttp1OnlyHeaders(io.netty.handler.codec.http.HttpHeaders, HttpHeadersBuilder)}.
      */
```
Note that I haven't migrated this method to use `purgeHttp1OnlyHeaders` since this is probably slower for other code paths using this method without zero-copy

jrhee17 reviewed (approved):
I believe this PR is ready for review by other maintainers - changes look good to me for merging

minwoox reviewed (approved):
I've also put a patch that
- hides static methods in ArmeriaHttpUtil
- changes ArmeriaHttpHeadersFactory to enum

Thanks!

ikhoon reviewed (approved):
Thanks, @echo304!
I'm looking forward to seeing benchmark results. What do you think of sending a PR to https://github.com/TechEmpower/FrameworkBenchmarks/blob/master/frameworks/Java/armeria/pom.xml#L19 when a new version is released?

jrhee17 commented:
I ran `com.linecorp.armeria.core.HttpServerBenchmark.plainText` a couple times

```sh
./gradlew :benchmarks:jmh:jmh \
      -Pjmh.includes='com.linecorp.armeria.core.HttpServerBenchmark.plainText' -Pjmh.params='protocol=H1C' -Pjmh.profilers="async:libPath=$HOME/Projects/async-profiler/build/lib/libasyncProfiler.dylib;output=flamegraph;dir=$HOME/result"
```

base
---
Benchmark                            (chunkCount)  (protocol)   Mode  Cnt      Score      Error  Units
HttpServerBenchmark.plainText                 100         H1C  thrpt    5  59030.256 ± 4749.908  ops/s
HttpServerBenchmark.plainText:async           100         H1C  thrpt             NaN               ---

feat
---
Benchmark                            (chunkCount)  (protocol)   Mode  Cnt      Score      Error  Units
HttpServerBenchmark.plainText                 100         H1C  thrpt    5  60595.652 ± 5476.196  ops/s
HttpServerBenchmark.plainText:async           100         H1C  thrpt             NaN               ---

I don't think there is a significant difference in the simple case. Having said this, Assuming there are more custom headers (which I think better represents the real world) I suppose the performance improvement is more obvious.
